frontend/app/model/MembershipCatalog.scala

```diff
@@ -0,0 +1,219 @@
+      val partnerM = paidTierPlanDetails(PaidTierPlan.monthly(Partner, Public), ids.partnerMonthly)
+      val partnerY = paidTierPlanDetails(PaidTierPlan.yearly(Partner, Public), ids.partnerYearly)
+      val patronM = paidTierPlanDetails(PaidTierPlan.monthly(Patron, Public), ids.patronMonthly)
+      val patronY = paidTierPlanDetails(PaidTierPlan.yearly(Patron, Public), ids.patronYearly)
+      val all = List(friend, staff, supporterM, supporterY, partnerM, partnerY, patronM, patronY)
+    }
+
+    object LegacyPlans {
+      val ids = ratePlanIds.legacy
+      val friend = freeTierPlanDetails(FriendTierPlan.legacy, ids.friend)
+      val supporterM = paidTierPlanDetails(PaidTierPlan.monthly(Supporter, Legacy), ids.supporterMonthly)
+      val supporterY = paidTierPlanDetails(PaidTierPlan.yearly(Supporter, Legacy), ids.supporterYearly)
+      val partnerM = paidTierPlanDetails(PaidTierPlan.monthly(Partner, Legacy), ids.partnerMonthly)
+      val partnerY = paidTierPlanDetails(PaidTierPlan.yearly(Partner, Legacy), ids.partnerYearly)
+      val patronM = paidTierPlanDetails(PaidTierPlan.monthly(Patron, Legacy), ids.patronMonthly)
+      val patronY = paidTierPlanDetails(PaidTierPlan.yearly(Patron, Legacy), ids.patronYearly)
+      val all = List(friend, supporterM, supporterY, partnerM, partnerY, patronM, patronY)
+    }
+
+    val tierPlanDetailsMap: Val[Map[ProductRatePlanId, TierPlanDetails]] =
+      //The failed public plans are discarded as they are going to be accumulated while the tier details are computed
+      Applicative[Val].sequence(PublicPlans.all.filter(_.isSuccess) ++ LegacyPlans.all)
+        .map(_.map { d => d.productRatePlanId -> d }.toMap)
+
+    def paidTierDetails(monthly: Val[PaidTierPlanDetails], yearly: Val[PaidTierPlanDetails]): Val[PaidTierDetails] =
+      (monthly |@| yearly)(PaidTierDetails).assert(currenciesMismatchErr)(currenciesMatch)
+
+    object Details {
+      private val P = PublicPlans
+      val friend = P.friend.map(FreeTierDetails)
+      val staff = P.staff.map(FreeTierDetails)
+      val supporter = paidTierDetails(P.supporterM, P.supporterY)
+      val partner = paidTierDetails(P.partnerM, P.partnerY)
+      val patron = paidTierDetails(P.patronM, P.patronY)
+    }
+
+    ( Details.friend |@| Details.staff |@| Details.supporter |@| Details.partner |@| Details.patron |@|
+      tierPlanDetailsMap |@| bt.successNel
+    )(MembershipCatalog.apply)
+  }
+
+  def unsafeFromZuora(ratePlanIds: RatePlanIds)(ratePlans: Seq[rest.ProductRatePlan])(implicit bt: BackendType): MembershipCatalog = {
+    fromZuora(ratePlanIds)(ratePlans) match {
+      case Success(catalog) => catalog
+      case Failure(errs) =>
+        val msg = s"The catalog is inconsistent:\n" + errs.list.map(errorLine).mkString("\n")
+        throw MembershipCatalogException(msg)
+    }
+  }
+
+  def errorLine(p: (TierPlan, String)) = s"  - ${p._1}: ${p._2}"
+}
+
+trait MembershipCatalog {
+  def friend: FreeTierDetails
```

Comment:
Please try `lazy vals` instead of `def` - these are supposed to be immutable values!